### ISSUE TITLE
google_checks: empty pair of curly braces (empty block) is concise and should not be violation for non multi-block statement 

### ISSUE THREAD

**Issue by Zopsss** [checkstyle/checkstyle#15338]
I have read documentation: https://checkstyle.org/styleguides/google-java-style-20220203/javaguide.html#s4.1.3-braces-empty-blocks
https://checkstyle.org/google_style.html#a4.1.3
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

Detected at: https://github.com/checkstyle/checkstyle/pull/15331#pullrequestreview-2190376287

google_checks.xml gives violation for having a pair of curly braces without containing whitespace between them ( `{}` ) for `if`, `try`, etc statements, for example:

```java
/** some. */
public class Test {
  void testing() throws Exception {
    int xyz = 10;

    // 2 violations 3 lines below:
    //  '{' is not followed by whitespace
    //  '}' is not preceded with whitespace
    try (AutoCloseable a = null) {}

    // 2 violations 3 lines below:
    //  '{' is not followed by whitespace
    //  '}' is not preceded with whitespace
    if (xyz == 45) {}

    // 2 violations 3 lines below:
    //  '{' is not followed by whitespace
    //  '}' is not preceded with whitespace
    try (AutoCloseable a = null) {}

    // 2 violations 3 lines below:
    //  '{' is not followed by whitespace
    //  '}' is not preceded with whitespace
    if (xyz == 45) {}

    testing2();
  }

  void testing2() {}

  void testing3() {}

  class InnerClass {}

  class InnerClass2 {}

  interface InnerInterface {}

  interface InnerInterface2 {}

  enum InnerEnum {}

  enum InnerEnum2 {}

  record InnerRecord(int abc) {}

  record InnerRecord2(int abc) {}

  interface InnerInterface3 {
    static void foo() {}

    static void foo2() {}
  }

  abstract class InnerAbstractClass {
    void foo() {}

    void foo2() {}
  }

  static {
  } // violation is detected by Style guide allows this.

  static {
  }
}
```

**Output:**
```
$ java -jar checkstyle-10.18.1-all.jar -c /google_checks.xml Test.java | grep -v "EmptyBlock"
Starting audit...
[WARN] /var/tmp/Test.java:9:34: WhitespaceAround: '{' is not followed by whitespace.
 Empty blocks                may only be represented as {} when
 not part of a multi-block statement (4.1.3) [WhitespaceAround]
[WARN] /var/tmp/Test.java:9:35:
WhitespaceAround: '}' is not preceded with whitespace. [WhitespaceAround]

[WARN] /var/tmp/Test.java:14:20: WhitespaceAround: '{' is not followed by whitespace. Empty blocks                may only be represented as {} when not part of a multi-block statement (4.1.3) [WhitespaceAround]
[WARN] /var/tmp/Test.java:14:21: WhitespaceAround: '}' is not preceded with whitespace. [WhitespaceAround]
[WARN] /var/tmp/Test.java:19:34: WhitespaceAround: '{' is not followed by whitespace. Empty blocks                may only be represented as {} when not part of a multi-block statement (4.1.3) [WhitespaceAround]
[WARN] /var/tmp/Test.java:19:35: WhitespaceAround: '}' is not preceded with whitespace. [WhitespaceAround]
[WARN] /var/tmp/Test.java:24:20: WhitespaceAround: '{' is not followed by whitespace. Empty blocks                may only be represented as {} when not part of a multi-block statement (4.1.3) [WhitespaceAround]
[WARN] /var/tmp/Test.java:24:21: WhitespaceAround: '}' is not preceded with whitespace. [WhitespaceAround]
Audit done.
```

Config is complaining about not having a whitespace between curly braces for lines:

`try (AutoCloseable a = null) {}` & `if (xyz == 45) {}`

Config is only complaining for `try` and `if` statements, it is not complaining for other code which also uses `{}`.

It is not mentioned anywhere in the style guide that an empty pair of curly braces should have whitespace between them, so this might be a false positive in our config. Having `{}` types of braces is ok, instead we should give violation for `{ }` types of braces.

The google-java-format tool also removes whitespaces from such empty curly braces ( `{ }` ) and formats them to something like: `{}`

---

Braces rules are at https://google.github.io/styleguide/javaguide.html#s4.1.2-blocks-k-r-style
according to this section they should be on formatted as formatter is doing. Behavior of formatter is ok to make:
```
static {
}
```

There is one more section https://google.github.io/styleguide/javaguide.html#s4.1.3-braces-empty-blocks it makes allowance for `{}`.
> An empty block or block-like construct may be in K & R style (as described in [Section 4.1.2](https://google.github.io/styleguide/javaguide.html#s4.1.2-blocks-k-r-style)). Alternatively, it may be closed immediately after it is opened, with no characters or line break in between ({}), unless it is part of a multi-block statement (one that directly contains multiple blocks: if/else or try/catch/finally).

Attention to `with no characters or line break in between ({})`.

So code `static {}` is valid by Style guide.
WhitespaceAround config in google_checks.xml should be somehow updated to avoid violation in such code.

Note: such code is nonsense , nobody writes like this, so priority of fix is low.

-----

**Describe what you expect in detail.**

Config should give violation `{ }` types of braces. It should not give violation for `{}` types of braces.

**Comment by rdiachenko** [comment]
> unless it is part of a multi-block statement (one that directly contains multiple blocks: if/else or try/catch/finally)

We should also cover these cases:
```java
if (condition1) {
  foo();
} else if (condition2) {} // violation because this is a part of multi-block statement

if (condition1) {
  foo();
} else {} // violation because this is a part of multi-block statement

try {
  foo();
} catch (RuntimeException e) {} // violation because this is a part of multi-block statement

try {
  foo();
} finally {} // violation because this is a part of multi-block statement
```

**Comment by rdiachenko** [comment]
> [4.1.3 Empty blocks: may be concise](https://google.github.io/styleguide/javaguide.html#s4.1.3-braces-empty-blocks)
> 
> An empty block or block-like construct may be in K & R style (as described in [Section 4.1.2](https://google.github.io/styleguide/javaguide.html#s4.1.2-blocks-k-r-style)). Alternatively, it may be closed immediately after it is opened, with no characters or line break in between ({}), unless it is part of a multi-block statement (one that directly contains multiple blocks: if/else or try/catch/finally).

After thinking more on this I'm not sure we have a correct configuration for this section.

> An empty block or block-like construct may be in K & R style

This means the following empty block is allowed:
```
static {
}
```

> Alternatively, it may be closed immediately after it is opened, with no characters or line break in between ({})

The main word here is **Alternatively** which means this empty block is also valid:
```
static {}
```

Google formatter formats the above block to be in K & R style which is fine.

On the Checkstyle side we use `EmptyBlock` and `EmptyCatchBlock` to cover 4.1.3 section. I think this doesn't make sense, as those checks complain for empty blocks presense, while the section says that it is ok to have empty blocks.

What we need to do is the following:
- do not give violations for empty blocks, don't use `EmptyBlock` and `EmptyCatchBlock` and find more suitable checks to cover this section
- if the empty block in K & R style, nothing to do here
- if the empty block in the alternative style (both `{` and `}` are on the same line):
  - for NON multi-block statements give violation only if there are whitespace characters in between `{` and `}` (e.g., `{   }`)
  - for multi-block statements always give violation, e.g.:
```
        if (a == -1) { } // violation, whitespace in between {}

        if (a == 0) {} // ok, no characters in between {} and not part of a multi-block statement

        if (a == 1) { // ok, K & R style
        }

        if (a == 2) { // ok, K & R style
        } else {} // violation, part of a multi-block statement

        // ok, K & R style
        if (a == 3) {
        } else {
        }
```

**Comment by romani** [comment]
>We should also cover these cases:

Such cases are part of multi block, there should be some content in `{}`. So violations valid.

**Comment by romani** [comment]
@rdiachenko , 
https://github.com/checkstyle/checkstyle/blob/a26873aee09926ac6e584af67dd9ba8cd069fdc9/src/main/resources/google_checks.xml#L125-L133

this is how we already allow `{}` for methods and any other strucutures.
May be we should just extend this list of properties or make some more general property?

**Comment by romani** [comment]
match empty static block:
```
$ cat Test.java 
/** some. */
public class Test {

  static {}

  static {;}

}
$ java -jar checkstyle-10.18.1-all.jar \
  -b "//STATIC_INIT/SLIST/*[1][self::RCURLY]" \
  Test.java
COMPILATION_UNIT -> COMPILATION_UNIT [2:0]
`--CLASS_DEF -> CLASS_DEF [2:0]
    `--OBJBLOCK -> OBJBLOCK [2:18]
        |--STATIC_INIT -> STATIC_INIT [4:2]
        |   `--SLIST -> { [4:9]
        |       `--RCURLY -> } [4:10]

```

**Comment by romani** [comment]
empty IF
```
$ javac Test.java 
$ cat Test.java 
/** some. */
public class Test {

  void foo() {
    if (true) { }
    if (true) {;}
  }

}
$ java -jar checkstyle-10.18.1-all.jar -b "//LITERAL_IF/SLIST/*[1][self::RCURLY]" Test.java
COMPILATION_UNIT -> COMPILATION_UNIT [2:0]
`--CLASS_DEF -> CLASS_DEF [2:0]
    `--OBJBLOCK -> OBJBLOCK [2:18]
        |--METHOD_DEF -> METHOD_DEF [4:2]
        |   `--SLIST -> { [4:13]
        |       |--LITERAL_IF -> if [5:4]
        |       |   `--SLIST -> { [5:14]
        |       |       `--RCURLY -> } [5:16]

```

**Comment by romani** [comment]
`//*[self::LITERAL_IF or self::STATIC_INIT or self::LITERAL_TRY]/SLIST/*[1][self::RCURLY]`

**Comment by rdiachenko** [comment]
We still need to cover multi-block statements:
```java
        if (a == 2) { // ok, K & R style
        } else {} // violation, part of a multi-block statement

        // ok, K & R style
        if (a == 3) {
        } else {
        }
```
`RightCurly` might help, but I haven't checked it.

**Comment by romani** [comment]
```
$ diff -u google_checks.xml google_checks_updated.xml 
--- google_checks.xml	2024-09-11 05:33:21.000000000 -0700
+++ google_checks_updated.xml	2024-09-11 05:57:35.806936310 -0700
@@ -78,11 +78,6 @@
     <module name="NoLineWrap">
       <property name="tokens" value="PACKAGE_DEF, IMPORT, STATIC_IMPORT"/>
     </module>
-    <module name="EmptyBlock">
-      <property name="option" value="TEXT"/>
-      <property name="tokens"
-               value="LITERAL_TRY, LITERAL_FINALLY, LITERAL_IF, LITERAL_ELSE, LITERAL_SWITCH"/>
-    </module>
     <module name="NeedBraces">
       <property name="tokens"
                value="LITERAL_DO, LITERAL_ELSE, LITERAL_FOR, LITERAL_IF, LITERAL_WHILE"/>
@@ -144,6 +139,14 @@
       <message key="ws.notPreceded"
               value="WhitespaceAround: ''{0}'' is not preceded with whitespace."/>
     </module>
+    <module name="SuppressionXpathSingleFilter">
+      <property name="checks" value="WhitespaceAround"/>
+      <property name="query" value="//*[self::LITERAL_IF or self::LITERAL_ELSE or self::STATIC_INIT or self::LITERAL_TRY]/SLIST[count(./*)=1]"/>
+    </module>
+    <module name="RegexpSinglelineJava">
+      <property name="format" value="\{[ ]+\}"/>
+      <property name="message" value="Empty blocks should have no spaces."/>
+    </module>
     <module name="OneStatementPerLine"/>
     <module name="MultipleVariableDeclarations"/>
     <module name="ArrayTypeStyle"/>

$ cat Test.java 
class Test {
  void foo() {
    // ok, K & R style
    if (a == 2) {
    } else {} // violation, part of a multi-block statement

    // ok, K & R style
    if (a == 2) {
    } else { } // violation, part of a multi-block statement

    // ok, K & R style
    if (a == 3) {
    } else {
    }
  }
}

$ java -jar checkstyle-10.18.1-all.jar -c google_checks_updated.xml Test.java
Starting audit...
[WARN] /var/tmp/Test.java:5:13: WhitespaceAround: '}' is not preceded with whitespace. [WhitespaceAround]
[WARN] /var/tmp/Test.java:9: Empty blocks should have no spaces. [RegexpSinglelineJava]
Audit done.

```

**Comment by rdiachenko** [comment]
We may need to also remove `EmptyCatchBlock` for this section and extend the xpath for `try-catch-finally` multi-blocks. `switch-case-default` is another multi-block statement I can think of. But I agree we can start with your suggestion for `static` and `if-else` blocks, see how it works, and extend same solution for the rest of the multi-blocks if that works.

**Comment by Zopsss** [comment]
> We may need to also remove `EmptyCatchBlock`

We cannot remove this module as it is required by section [6.2](https://checkstyle.org/google_style.html#a6.2)

**Comment by Zopsss** [comment]
>   void foo() {
    // ok, IF braces are K & R style
    if (a == 2) {
    } else {} // violation, 'WhitespaceAround: '}' is not preceded with whitespace'

We should be giving violation for not having curly braces in K&R style for this case, formatter also formatted it to be K&R style:

```
$ cat TestingCurlyBrackets.java 
public class TestingCurlyBrackets {
  void foo() {
    // ok, IF braces are K & R style
    if (a == 2) {
    } else {} // violation, 'WhitespaceAround: '}' is not preceded with whitespace'
  }
}

$ java -jar google-java-format-1.23.0-all-deps.jar --replace TestingCurlyBrackets.java 

$ cat TestingCurlyBrackets.java
public class TestingCurlyBrackets {
  void foo() {
    // ok, IF braces are K & R style
    if (a == 2) {
    } else {
    } // violation, 'WhitespaceAround: '}' is not preceded with whitespace'
  }
}
```

**Comment by romani** [comment]
> We should be giving violation for not having curly braces in K&R style for this case,

We can split issue in few issues, in first issue we just remove all false-positives, so it means bad violations. It can annoy users as they can not avoid/remove it.

In next issue we will try to handle false-negatives, so it means - there is no violations and but it should. Users are not blocked (by CI build for example) if there is missed violation. Not ideal situation but not blocked.

**Comment by Zopsss** [comment]
> > We should be giving violation for not having curly braces in K&R style for this case,
> 
> We can split issue in few issues, in first issue we just remove all false-positives, so it means bad violations. It can annoy users as they can not avoid/remove it.
> 
> In next issue we will try to handle false-negatives, so it means - there is no violations and but it should. Users are not blocked (by CI build for example) if there is missed violation. Not ideal situation but not blocked.

Sounds like a good plan to me, hope that it'll not take more to finish


> Issue is moved to TODO state, @Zopsss  please start updating of config, testing, and turning of config to make it work.


I'm waiting for https://github.com/checkstyle/checkstyle/pull/15655 to get merged, after that I'll start updating config

**Comment by romani** [comment]
No need to wait, just make new branch from branch of this PR, do work, you can send PR (there will be 2 commits), we start review  of second commit, you can do testing by diff report,..... . After #15655 merged, you will just rebase to avoid first commit.

**Comment by Zopsss** [comment]
> We can split issue in few issues, in first issue we just remove all false-positives, so it means bad violations. It can annoy users as they can not avoid/remove it.

an issue for this has been opened: https://github.com/checkstyle/checkstyle/issues/15658 @romani
